src/watchdog/src/watchdog.py

```diff
@@ -237,44 +244,59 @@ def parse_pod_item(pod, pai_pod_gauge, pai_container_gauge, pods_info):
                 error_counter.labels(type="unknown_pod_cond").inc()
                 logger.warning("unexpected condition %s in pod %s", cond_t, pod_name)
 
-    pai_pod_gauge.add_metric([service_name, pod_name, namespace, phase, host_ip,
-        initialized, pod_scheduled, ready], 1)
+    if service_name is not None:
+        pai_pod_gauge.add_metric([service_name, pod_name, namespace, phase, host_ip,
+            initialized, pod_scheduled, ready], 1)
+    if job_name is not None:
+        pai_job_pod_gauge.add_metric([job_name, pod_name, phase, host_ip,
+            initialized, pod_scheduled, ready], 1)
 
-    # generate pai_containers
-    if status.get("containerStatuses") is not None:
-        container_statuses = status["containerStatuses"]
 
-        for container_status in container_statuses:
-            container_name = container_status["name"]
+def generate_pods_info(pod_name, containers, host_ip, pods_info):
+    used_gpu = 0
+    if containers is not None:
+        for container in containers:
+            req_gpu = int(walk_json_field_safe(container, "resources", "requests",
+                    "nvidia.com/gpu") or 0)
+            limit_gpu = int(walk_json_field_safe(container, "resources", "limits",
+                    "nvidia.com/gpu") or 0)
+            used_gpu += max(req_gpu, limit_gpu)
+    pods_info[host_ip].append(PodInfo(pod_name, used_gpu))
 
-            ready = False
 
-            if container_status.get("ready") is not None:
-                ready = container_status["ready"]
+def generate_container_metrics(pai_container_gauge, service_name, pod_name, container_statuses,
+                               namespace, host_ip):
+    for container_status in container_statuses:
+        container_name = container_status["name"]
 
-            container_state = None
-            if container_status.get("state") is not None:
-                state = container_status["state"]
-                if len(state) != 1:
-                    error_counter.labels(type="unexpected_container_state").inc()
-                    logger.error("unexpected state %s in container %s",
-                            json.dumps(state), container_name)
-                else:
-                    container_state = list(state.keys())[0].lower()
+        ready = False
 
-            pai_container_gauge.add_metric([service_name, pod_name, container_name,
-                namespace, container_state, host_ip, str(ready).lower()], 1)
+        if container_status.get("ready") is not None:
```

Comment:
use Pod.Phase = Pending to judge this

Reply:
Sure, we use `Pod.Status.Phase` to judge this. Please refer to [https://github.com/microsoft/pai/blob/5ecde18ae3febb3e66f9c40cfccd85a483eca585/src/prometheus/deploy/alerting/jobs.rules#L29](https://github.com/microsoft/pai/blob/5ecde18ae3febb3e66f9c40cfccd85a483eca585/src/prometheus/deploy/alerting/jobs.rules#L29) and [https://github.com/microsoft/pai/blob/5ecde18ae3febb3e66f9c40cfccd85a483eca585/src/watchdog/src/watchdog.py#L224-L251](https://github.com/microsoft/pai/blob/5ecde18ae3febb3e66f9c40cfccd85a483eca585/src/watchdog/src/watchdog.py#L224-L251). Watchdog just update pod status to promethus. Then we can define promethus rules for alerts

---
In reply to: [337321537](https://github.com/microsoft/pai/pull/3761#discussion_r337321537) [](ancestors = 337321537)